### PR TITLE
fix(planner): make sure schema is consistant between logical and physical plan

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #issue

### CODE REVIEW

File: tests/sqllogictests/suites/query/union.test [hunk 1]
```diff
@@ -177,10 +177,3 @@ drop table t1
 
 statement error 1065
 select [1,2,3] union all select 2
-
-# type coercion
```
Comment (xudong963) [inline tests/sqllogictests/suites/query/union.test:181]: Why delete the test?

Comment (andylokandy) [reply tests/sqllogictests/suites/query/union.test:181]: Oh, I overlooked the result of the git revert.

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 I got it. Here's the human readable git commit message:

Fix type mismatch issue in HashJoin and push down predicates to plan nodes

This patch fixes two issues in the SQL executor:

1. Handle type mismatches in HashJoin keys more rigorously. Previously we did some simple type casting which is not sufficient. This patch unifies the types of left and right exchange keys properly by finding their common super type and performing explicit casts.

2. Fix the logic of pushing down predicates to plan nodes. Previously we only did simple type checking and casting, which leads to failures. This patch fixes it by finding the common super type of expressions involved in a predicate and performing appropriate casts.

This ensures that predicates are pushed down correctly to plan nodes, and join keys have consistent types. As a result, queries with type mismatches in join keys or predicates will now work correctly.

The key changes are:

- Rewrite HashJoin builder to properly unify the types of left and right exchange keys.
- Handle type mismatches when pushing down predicates to joins, scans, EvalScalars, etc. 
- Unify the types of ORDER BY expressions and RANGE offsets when resolving window frames.
- Perform explicit type casts when decorrelating subqueries.

Overall this patch improves the type checking and type casting logic throughout the executor, resulting in more robust query execution.

Hope this helps! Let me know if you have any other questions.

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 I commit message:

Improved type consistency for logical plans

This commit improves data type consistency for some logical plans:

1. HashJoin, RangeJoin: Any data types mismatch between join keys will be coerced to a super type automatically. This avoids unnecessary casting in PhysicalPlanBuilder.

2. Window: 

- Any data types mismatch between partition by/order by columns and WINDOW function arguments will be coerced to a super type. 
- For RANGE frames, data types between ORDER BY and RANGE offset will also be coerced to a super type.
- This allows us to build the Window plan without casting in PhysicalPlanBuilder.

3. Subquery: For correlated subqueries, any data type mismatch between the subquery output and the comparison expression will be coerced to a super type. This avoids unnecessary casting in PhysicalPlanBuilder.

4. Filter predicates: Any  data type mismatch between Filter predicates and other logical plans will be handled during rule transformation, instead of PhysicalPlanBuilder. This improves data type consistency.

Overall this commit improves data type consistency for logical plans by handling type casting and coercion during logical plan analysis and optimization. It simplifies PhysicalPlanBuilder by moving type related logic into logical plan binder.

Hope this helps! Let me know if you have any questions.

**Comment by BohuTANG** [comment]
askbend:summary

**Pull request closed without merging**